File: android/app/src/main/java/com/example/wavify/SttEngine.java
package com.example.wavify;

import android.content.Context;

public class SttEngine {
    private Context context;
    public SttEngine(Context context) {
        this.context = context;
    }

    private static native long createFfi(final String modelPath, final String apiKey, final String appName);
    public long create(String modelPath, String apiKey) {
        String appName = context.getPackageName();
        return createFfi(modelPath, apiKey, appName);
    }

    private static native long destroyFfi(final long model);
    public long destroy(long model) {
        return destroyFfi(model);
    }

    private static native String sttFfi(final float[] data, final long model);
    public String stt(float[] data, long model) {return sttFfi(data, model);};

}

File: android/wavify/src/main/java/dev/wavify/SttEngine.kt
package dev.wavify

import android.util.Log

enum class LogLevel(val level: String) {
    TRACE("trace"),
    DEBUG("debug"),
    INFO("info"),
    WARN("warn"),
    ERROR("error")
}

object Logger {
    init {
        System.loadLibrary("wavify_core")
        System.loadLibrary("tensorflowlite_c")
    }

    @JvmStatic
    private external fun setupLoggerFfi(logLevel: String)

    fun setLogLevel(logLevel: LogLevel = LogLevel.INFO) {
        setupLoggerFfi(logLevel.level)
    }
}

class SttEngine private constructor(private val nativeHandle: Long) {

    companion object {

        val TAG: String = "Wavify"

        @JvmStatic
        external fun createFfi(modelPath: String, apiKey: String, appName: String): Long

        @JvmStatic
        external fun destroyFfi(engineHandle: Long)

        @JvmStatic
        external fun sttFfi(data: FloatArray, engineHandle: Long): String

        fun create(modelPath: String, apiKey: String, appName: String): SttEngine {

            Log.d(TAG,"linking libraries")
            System.loadLibrary("wavify_core")
            System.loadLibrary("tensorflowlite_c")

            Log.d(TAG, "initializing engine")
            val handle = createFfi(modelPath, apiKey, appName)
            return SttEngine(handle)
        }
    }

    fun destroy() {
        destroyFfi(nativeHandle)
    }

    fun stt(data: FloatArray): String {
        return sttFfi(data, nativeHandle)
    }
}


